sbnanalysis/ana/SBNOsc/NumuSelection.cxx

```diff
@@ -37,31 +145,228 @@ bool NumuSelection::ProcessEvent(const gallery::Event& ev, std::vector<Event::In
+        _root_histos[select_i].h_numu_contained_L->Fill(intInfo.l_contained_length);
+        _root_histos[select_i].h_numu_l_is_contained->Fill(intInfo.l_is_contained);
+        _root_histos[select_i].h_numu_Vxy->Fill(nu.Nu().Vx(), nu.Nu().Vy());
+        _root_histos[select_i].h_numu_Vxz->Fill(nu.Nu().Vx(), nu.Nu().Vz());
+        _root_histos[select_i].h_numu_Vyz->Fill(nu.Nu().Vy(), nu.Nu().Vz());
+      }
+      select_i ++;
+    }
+  }
+  return selected;
+}
+
+NumuSelection::NuMuInteraction NumuSelection::interactionInfo(const gallery::Event &ev, const simb::MCTruth &mctruth) {
+  // get handle to tracks and showers
+  auto const& mctrack_list = \
+    *ev.getValidHandle<std::vector<sim::MCTrack> >(_config.mctTag);
+  auto const& mcshower_list = \
+    *ev.getValidHandle<std::vector<sim::MCShower> >(_config.mcsTag);
+
+  // and particles
+  auto const& mcparticle_list = \
+    *ev.getValidHandle<std::vector<simb::MCParticle>>(_config.mcpTag);
+
+  // Get the length and determine if any point leaves the fiducial volume
+  //
+  // get lepton track
+  int lepton_ind = -1;
+  for (int i = 0; i < mctrack_list.size(); i++) {
+    if (isFromNuVertex(mctruth, mctrack_list[i]) && mctrack_list[i].PdgCode() == 13 && mctrack_list[i].size() > 0) {
+      lepton_ind = i;
+      break;
+    }
+  } 
+
+  // parameters only make sense if lepton exists
+  bool contained_in_FV = false;
+  double l_contained_length = -1;
+  double l_length = -1;
+  double smeared_eccqe = -1;
+
+  if (lepton_ind != -1) {
+    // if lepton exists, addume contained by default
+    contained_in_FV = true;
+    l_contained_length = 0;
+    l_length = 0;
+    smeared_eccqe = 0;
+
+    auto const& lepton_track = mctrack_list.at(lepton_ind);
+    
+    // Get the length and determine if any point leaves the fiducial volume
+    TLorentzVector pos = lepton_track.Start().Position();
+    for (int i = 1; i < lepton_track.size(); i++) {
+      // update if track is contained
+      if (contained_in_FV) contained_in_FV = containedInFV(pos.Vect());
+      
+      // update length
+      l_contained_length += containedLength(lepton_track[i].Position().Vect(), pos.Vect(), _config.fiducial_volumes);
+      l_length += (lepton_track[i].Position().Vect() - pos.Vect()).Mag();
+      
+      pos = lepton_track[i].Position();
     }
+    
+    // determine if stopped in TPC
+    // sim::MCStep end = lepton_track.End();
+    // geoalgo::Point_t l_end_pos(end.Position().Vect());
+    // bool stop_in_tpc = _config.active_volume.Contain(l_end_pos);
+
+    // While we're here, get the smeared ECCQE
+    smeared_eccqe = ECCQE(lepton_track.Start().Momentum().Vect(), lepton_track.Start().E()/1000., 
+      _smear->Smear(lepton_track.Start().E()/1000., 14, contained_in_FV));
   }
+    
+  // get visible energy
+  // "Reconstruct" visible energy from tracks + showers
+  // Also do smearing
+  //
+  // Track energy and masses are in MeV
+  double visible_E = 0.;
```

Comment:
I think this may need to be refined a bit... it looks like this could include not-so-visible energy from neutral particles, and isn't really a true visible energy since it doesn't account for containment or energy/tracking threshold.

Reply:
Yeah -- I am not positive about the visible energy definition.

I have moved the value into the "Event" class so that it can be used downstream. I have also removed neutral particles from the calculation. However, I don't think the proposal analysis accounted for containment or thresholds.

Reply:
It does account for thresholds actually, so that needs to be added in.

Reply:
Added in a first pass of threshold.